test/framework/src/main/java/org/elasticsearch/test/disruption/LongGCDisruption.java

```diff
@@ -49,13 +51,66 @@ public LongGCDisruption(Random random, String disruptedNode) {
     @Override
     public synchronized void startDisrupting() {
         if (suspendedThreads == null) {
-            suspendedThreads = new HashSet<>();
-            stopNodeThreads(disruptedNode, suspendedThreads);
+            boolean success = false;
+            try {
+                suspendedThreads = ConcurrentHashMap.newKeySet();
+
+                final String currentThreadName = Thread.currentThread().getName();
+                assert currentThreadName.contains("[" + disruptedNode + "]") == false :
+                    "current thread match pattern. thread name: " + currentThreadName + ", node: " + disruptedNode;
+                // we spawn a background thread to protect against deadlock which can happen
+                // if there are shared resources between caller thread and and suspended threads
+                // see unsafeClasses to how to avoid that
+                final AtomicReference<Exception> stoppingError = new AtomicReference<>();
+                final Thread stoppingThread = new Thread(new AbstractRunnable() {
+                    @Override
+                    public void onFailure(Exception e) {
+                        stoppingError.set(e);
+                    }
+
+                    @Override
+                    protected void doRun() throws Exception {
+                        while (stopNodeThreads(disruptedNode, suspendedThreads)) ;
+                    }
+                });
+                stoppingThread.setName(currentThreadName + "[LongGCDisruption][threadStopper]");
+                stoppingThread.start();
+                try {
+                    stoppingThread.join(getStoppingTimeoutInMillis());
+                } catch (InterruptedException e) {
+                    Thread.currentThread().interrupt();
+                }
+                if (stoppingError.get() != null) {
+                    throw new RuntimeException("unknown error while stopping threads", stoppingError.get());
+                }
+                if (stoppingThread.isAlive()) {
+                    logger.warn("failed to stop node [{}]'s threads within [{}] millis. Stopping thread stack trace:\n {}"
+                        , disruptedNode, getStoppingTimeoutInMillis(), stackTrace(stoppingThread));
+                    stoppingThread.interrupt(); // best effort;
```

Comment:
I don't see where the interrupt flag is ever checked on the stopping thread, what am I missing?